torch/lib/c10d/ProcessGroupWrapper.cpp

```diff
@@ -23,46 +23,68 @@ struct CollectiveFingerPrint {
   // Ref to input tensors, if given, of the collective. If given, shapes will be
   // checked across processes to ensure valid input into the collective.
   const std::vector<at::Tensor>& input_tensors_;
+  // input tensor data types
+  std::vector<int> tensor_dtypes_;
+  // input tensor device types
+  std::vector<int8_t> tensor_device_types_;
```

Comment:
nit: maybe use c10::DeviceType and c10::ScalarType here?

Reply:
The reason it is int instead of c10::DeviceType and c10::ScalarType is because we need to push them into a `int64` vector, and call allgather on that. Later if there is an error, we convert back to DeviceType/ScalarType as needed when logging.

So since we will have to do the conversion at some point, we can just store it as int, or do the conversion when pushing back to the allgather tensor. Either approach sounds good to me, do you have any thoughts?